src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -43,6 +45,153 @@ void Bootstrap(const MatType& dataset,
     bootstrapWeights = weights.cols(indices);
 }
 
+/**
+ * Default bootstrap strategy that uses the ::Bootstrap() function. 
+ */
+template<bool UseWeights>
+class DefaultBootstrap
+{
+ public:
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    mlpack::Bootstrap<UseWeights>(dataset,
+                                  labels,
+                                  weights,
+                                  bootstrapDataset,
+                                  bootstrapLabels,
+                                  bootstrapWeights);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+template<bool UseWeights>
+class IdentityBootstrap
+{
+ public:
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    bootstrapDataset = dataset;
+    bootstrapLabels = labels;
+    if (UseWeights)
+      bootstrapWeights = weights;
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<bool UseWeights, typename G = std::mt19937>
+class SequentialBootstrap
+{
+ public:
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               const G& generator = G(std::random_device()())) :
+    SequentialBootstrap(indicatorMatrix, indicatorMatrix.n_rows, generator)
+  {
+  }
+
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               arma::uword sampleCount,
+                               const G& generator = G(std::random_device()())):
+    indicatorMatrix(indicatorMatrix),
+    colIndices(arma::linspace<arma::uvec>(0, indicatorMatrix.n_cols - 1)),
+    sampleCount(sampleCount),
+    generator(generator)
+  {
+  }
+
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
```

Comment:
Note that `MatType` may not always be the same as `indicatorMatrix`.  So it may be better to have `indicatorMatrix` take the type `MatType` instead.  (This is much easier if we can avoid `indicatorMatrix` being a constructor argument.)